File: research/joinery/corpus/ledger.js
import { readFile } from "node:fs/promises";

export class Ledger {
  #rows = [];
  #total = null;

  constructor(seed = []) {
    for (const v of seed) this.push(v);
  }

  push(v) {
    if (typeof v !== "number") throw new TypeError(`not a number: ${v}`);
    this.#rows.push(v);
    this.#total = null;
    return this;
  }

  get total() {
    if (this.#total === null) {
      this.#total = this.#rows.reduce((a, r) => (r > 0 ? a + r : a), 0);
    }
    return this.#total;
  }

  static async load(path) {
    const text = await readFile(path, "utf8");
    return new Ledger(JSON.parse(text).rows ?? []);
  }
}

const led = new Ledger([1, 2, 3]);
console.log(`total=${led.total}`);

File: research/joinery/corpus/Ledger.java
package com.example;

import java.util.ArrayList;
import java.util.List;

public final class Ledger {
    private final List<Integer> rows = new ArrayList<>();
    private Integer total;

    public Ledger(List<Integer> seed) {
        if (seed != null) rows.addAll(seed);
    }

    public int total() {
        if (total == null) {
            int acc = 0;
            for (int r : rows) if (r > 0) acc += r;
            total = acc;
        }
        return total;
    }

    public Ledger merge(Ledger other) {
        rows.addAll(other.rows);
        total = null;
        return this;
    }

    public static void main(String[] args) {
        Ledger led = new Ledger(List.of(1, 2, 3));
        System.out.println("total=" + led.total());
    }
}
